pandas/core/dtypes/astype.py

```diff
@@ -100,6 +100,11 @@ def _astype_nansafe(
     elif np.issubdtype(arr.dtype, np.floating) and dtype.kind in "iu":
         return _astype_float_to_int_nansafe(arr, dtype, copy)
 
+    elif np.issubdtype(arr.dtype, np.complexfloating) and is_object_dtype(dtype):
+        if np.isnan(arr).any():
+            res = np.asarray([np.nan if np.isnan(x) else x for x in arr], dtype)
```

Comment:
Could you just do `res[np.isnan(arr)] = np.nan`?

Reply:
`res` has to be set to something first (it's the "result").  But I can look at using `copy` to decide if we really need something fresh before we do something that may have a side-effect.